File: src/cmcandy/solutions/_0023.java
package cmcandy.solutions;

/**
 * Created by CHEN on 2020/4/21.
 */
public class _0023 {

    /**
     * Definition for singly-linked list.
     * public class ListNode {
     *     int val;
     *     ListNode next;
     *     ListNode(int x) { val = x; }
     * }
     */
    static class ListNode {
        int val;
        ListNode next;
        ListNode(int x) {
            val = x;
        }
    }
    static class Solution {
        public ListNode mergeKLists(ListNode[] lists) {
            if(lists.length == 0)
                return null;
            if(lists.length == 1)
                return lists[0];
            if(lists.length == 2){
                return mergeTwoLists(lists[0],lists[1]);
            }
            int mid = lists.length/2;
            ListNode[] l1 = new ListNode[mid];
            ListNode[] l2 = new ListNode[lists.length- mid];
            for (int i = 0 ;i<mid;i++){
                l1[i] = lists[i];
            }
            for(int i = mid;i<lists.length;i++){
                l2[i-mid] = lists[i];
            }
            return mergeTwoLists(mergeKLists(l1),mergeKLists(l2));
        }

        public ListNode mergeTwoLists(ListNode l1,ListNode l2) {
            if (l1==null){
                return l2;
            }else if (l2==null){
                return l1;
            }
            ListNode head, tail;
            //第一个节点
            if (l1.val > l2.val) {
                head = l2;
                l2=l2.next;
            } else {
                head = l1;
                l1=l1.next;
            }
            tail = head;
            while (l1 != null || l2 != null) {
                if (l1 == null && l2 != null) {
                    tail.next = l2;
                    l2 = l2.next;
                } else if (l1 != null && l2 == null) {
                    tail.next = l1;
                    l1 = l1.next;
                }else if (l1.val > l2.val) {
                    tail.next = l2;
                    l2 = l2.next;
                }else if (l1.val <= l2.val) {
                    tail.next = l1;
                    l1 = l1.next;
                }
                tail = tail.next;
            }
            return head;
        }

    }



}

File: src/cmcandy/Python_language_Answers/_0023.py
from typing import List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        head = ListNode()
        if len(lists) == 0:
            return None
        now = head
        self.init = []
        for index in range(len(lists)):
            if lists[index] is not None:
                self.init.append([lists[index].val, index])
        nowIndex = 0
        length = len(self.init)
        # 优化：这里改成自定义排序就行了，类似于胜者树的思想吧，只不过改成维护一个数组
        self.init.sort()
        while nowIndex < length:
            now.next = lists[self.init[nowIndex][1]]
            now = now.next
            lists[self.init[nowIndex][1]] = lists[self.init[nowIndex][1]].next
            if lists[self.init[nowIndex][1]] is None:
                nowIndex += 1
            else:
                self.init[nowIndex][0] = lists[self.init[nowIndex][1]].val
            self.mySort(nowIndex, length)
        return head.next

    def mySort(self, start=0, end=0):
        '''
            维护队列用的sort主要是为了找到新的元素的位置,也可以用二分法找
            实际上胜者树可能更好，但是维护队列可能更简单
        '''
        if start == end:
            return
        tmp = self.init[start]
        for index in range(start + 1, end):
            if self.init[index][0] >= tmp[0]:
                self.init[index - 1] = tmp
                return
            else:
                self.init[index - 1] = self.init[index]
        self.init[end - 1] = tmp
        return
